src/algebra/parity.lean

```diff
@@ -8,50 +8,121 @@ import algebra.ring.basic
 import algebra.algebra.basic
 import algebra.group_power.basic
 import algebra.field_power
+import algebra.opposites
 
-/-!  This file proves some general facts about even and odd elements of semirings. -/
+/-!  # Squares, even and odd elements
 
+This file proves some general facts about squares, even and odd elements of semirings.
+
+In the implementation, we define `is_square` and we let `even` be the notion transported by
+`to_additive`.  The definition are therefore as follows:
+```lean
+is_square a ↔ ∃ r, a = r * r
+even a ↔ ∃ r, a = r + r
+```
+
+Odd elements are not unified with a multiplicative notion.
+
+* TODO: Try to generalize further the typeclass assumptions on `is_square/even`.
+  For instance, in some cases, there are `semiring` assumptions that I (DT) am not convinced are
+  necessary.
+* TODO: Consider moving the definition and lemmas about `odd` to a separate file.
+* TODO: The "old" definition of `even a` asked for the existence of an element `c` such that
+  `a = 2 * c`.  For this reason, several fixes introduce an extra `two_mul` or `← two_mul`.
+  It might be the case that by making a careful choice of `simp` lemma, this can be avoided.
+ -/
+
+open mul_opposite
 variables {α β : Type*}
 
-section semiring
-variables [semiring α] [semiring β] {m n : α}
+/--  An element `a` of a type `α` with multiplication satisfies `square a` if `a = r * r`,
+for some `r : α`. -/
+@[to_additive
```

Comment:
Which name does `to_additive` generate here? How does it know to turn `is_square` into `even`?

Reply:
This was a little naughty of me: I had already prepped `to_additive` to make the change.  [Here](https://github.com/leanprover-community/mathlib/blob/cdd1703cd342275133a47a4a88160ed16ec0a08e/src/algebra/group/to_additive.lean#L220) is the relevant line of code.